internal/server/v1/observations/derived_series.go

```diff
@@ -0,0 +1,507 @@
+	"sort"
+	"strings"
+
+	pb "github.com/datacommonsorg/mixer/internal/proto"
+	"github.com/datacommonsorg/mixer/internal/server/stat"
+	"github.com/datacommonsorg/mixer/internal/store"
+)
+
+// DerivedSeries implements API for Mixer.DerivedObservationsSeries.
+func DerivedSeries(
+	ctx context.Context,
+	in *pb.DerivedObservationsSeriesRequest,
+	store *store.Store,
+) (*pb.DerivedObservationsSeriesResponse, error) {
+	resp := &pb.DerivedObservationsSeriesResponse{}
+	entity := in.GetEntity()
+
+	// Parse the formula to extract all the variables, used for reading data from BT.
+	varCalculator, err := newVarCalculator(in.GetFormula())
+	if err != nil {
+		return resp, err
+	}
+	statVars := varCalculator.statVars()
+
+	// Read data from BT.
+	btData, err := stat.ReadStatsPb(
+		ctx, store.BtGroup, []string{entity}, statVars)
+	if err != nil {
+		return resp, err
+	}
+	entityData, ok := btData[entity]
+	if !ok {
+		return resp, err
+	}
+
+	// Calculate.
+	result, err := varCalculator.calculate(entityData)
+	if err != nil {
+		return resp, err
+	}
+	for _, p := range result.points {
+		resp.Observations = append(resp.Observations, &pb.PointStat{
+			Date:  p.GetDate(),
+			Value: p.GetValue(),
+		})
+	}
+
+	return resp, nil
+}
+
+type varCalculatorSeries struct {
+	statMetadata *pb.StatMetadata
+	// Sorted by date.
+	points []*pb.PointStat
+}
+
+// The date key of a time series is concatenation of all sorted dates.
+func (s *varCalculatorSeries) dateKey() string {
+	dates := []string{}
+	for _, point := range s.points {
+		dates = append(dates, point.GetDate())
+	}
+	return strings.Join(dates, "")
+}
+
+type varInfo struct {
+	statVar          string
+	statMetadata     *pb.StatMetadata
+	seriesCandidates []*varCalculatorSeries
+	chosenSeries     *varCalculatorSeries
+}
+
+type varCalculator struct {
+	expr ast.Expr
+	// Key is encodeForParse(varName).
+	varInfoMap map[string]*varInfo
+	debug      bool
+}
+
+func newVarCalculator(formula string) (*varCalculator, error) {
+	debug := os.Getenv("DEBUG") == "true"
+
+	expr, err := parser.ParseExpr(encodeForParse(formula))
+	if err != nil {
+		return nil, err
+	}
+
+	if debug {
+		fs := token.NewFileSet()
+		ast.Print(fs, expr)
+	}
+
+	// Iterate the AST, extract var names (variable with optional filters).
+	varNameVisitor := newVarNameVisitor()
+	ast.Walk(varNameVisitor, expr)
+
+	varInfoMap := map[string]*varInfo{}
+	for varName := range varNameVisitor.varNameSet {
+		varInfo, err := parseVarName(decodeForParse(varName))
+		if err != nil {
+			return nil, err
+		}
+		varInfoMap[varName] = varInfo
+	}
+
+	return &varCalculator{expr: expr, varInfoMap: varInfoMap, debug: debug}, nil
+}
+
+func (c *varCalculator) statVars() []string {
+	statVarSet := map[string]struct{}{}
+	for k := range c.varInfoMap {
+		statVarSet[c.varInfoMap[k].statVar] = struct{}{}
+	}
+	statVars := []string{}
+	for k := range statVarSet {
+		statVars = append(statVars, k)
+	}
+	return statVars
+}
+
+func (c *varCalculator) calculate(
```

Comment:
Instead of have a Calculator interface that calculate() a time series, I wonder if it's more flexible to just have a struct holding the "formula and operations" and have many functions to do the calculations like:

calculateSeries(f *Formula, s []*SourceSeries)
calculatePoints(f *Formula, s []*PointData)
calculateOtherSuff(...)

Now each calculator can only calculate one thing, which is effectively a function

Reply:
As discussed offline, the plan is to abstract out a lib just for AST parsing, which can be reused by ObsSeries, ObsColleciton, etc. This will be done in a future PR as this one is already complex.